heron/instance/src/java/com/twitter/heron/instance/HeronInstance.java

```diff
@@ -235,10 +235,13 @@ public void uncaughtException(Thread thread, Throwable exception) {
     private void handleException(Thread thread, Throwable exception) {
       // We would fail fast when errors occur to avoid unexpected bad situations
       if (exception instanceof Error) {
-        LOG.log(Level.SEVERE,
-            "Error caught in thread: " + thread.getName()
-                + " with thread id: " + thread.getId() + ". Process halting...");
-        Runtime.getRuntime().halt(1);
+        if (exception instanceof OutOfMemoryError) {
+          // Fail fast without logging, otherwise another OOM may happen
+          Runtime.getRuntime().halt(1);
```

Comment:
I guess you can do something like 
```
try {
  LOG.log("...")
} finally {
  Runtime.getRuntime().halt(1);
}
```

This both will log and most likely will halt process.